include/cert_generator.hpp

```diff
@@ -0,0 +1,691 @@
+}
+using EVP_PKEYPtr = openssl_ptr<EVP_PKEY, EVP_PKEY_free>;
+EVP_PKEYPtr makeEVPPKeyPtr(EVP_PKEY* ptr)
+{
+    return EVP_PKEYPtr(ptr, EVP_PKEY_free);
+}
+BIOPtr makeBIOPtr(BIO* ptr)
+{
+    return BIOPtr(ptr, BIO_free_all);
+}
+using EVP_PKEYPtr = openssl_ptr<EVP_PKEY, EVP_PKEY_free>;
+
+inline void printLastError()
+{
+    unsigned long err = ERR_get_error();
+    char err_buf[256];
+    ERR_error_string_n(err, err_buf, sizeof(err_buf));
+    LOG_ERROR("OpenSSL error: {}", err_buf);
+}
+// Convert PEM certificate file to DER format and save to file
+bool pemCertToDer(const std::string& pemPath, const std::string& derPath)
+{
+    FILE* pemFile = fopen(pemPath.c_str(), "r");
+    if (!pemFile)
+        return false;
+    X509Ptr cert =
+        makeX509Ptr(PEM_read_X509(pemFile, nullptr, nullptr, nullptr));
+    fclose(pemFile);
+    if (!cert)
+        return false;
+
+    FILE* derFile = fopen(derPath.c_str(), "wb");
+    if (!derFile)
+    {
+        return false;
+    }
+    int ret = i2d_X509_fp(derFile, cert.get());
+    fclose(derFile);
+    return ret == 1;
+}
+
+// Convert DER certificate file to PEM format and save to file
+bool derCertToPem(const std::string& derPath, const std::string& pemPath)
+{
+    FILE* derFile = fopen(derPath.c_str(), "rb");
+    if (!derFile)
+        return false;
+    X509Ptr cert = makeX509Ptr(d2i_X509_fp(derFile, nullptr));
+    fclose(derFile);
+    if (!cert)
+        return false;
+
+    FILE* pemFile = fopen(pemPath.c_str(), "w");
+    if (!pemFile)
+    {
+        return false;
+    }
+    int ret = PEM_write_X509(pemFile, cert.get());
+    fclose(pemFile);
+    return ret == 1;
+}
+
+// Convert PEM private key file to DER format and save to file
+bool pemKeyToDer(const std::string& pemPath, const std::string& derPath)
+{
+    FILE* pemFile = fopen(pemPath.c_str(), "r");
+    if (!pemFile)
+        return false;
+    EVP_PKEYPtr pkey =
+        makeEVPPKeyPtr(PEM_read_PrivateKey(pemFile, nullptr, nullptr, nullptr));
+    fclose(pemFile);
+    if (!pkey)
+        return false;
+
+    FILE* derFile = fopen(derPath.c_str(), "wb");
+    if (!derFile)
+    {
+        return false;
+    }
+    int ret = i2d_PrivateKey_fp(derFile, pkey.get());
+    fclose(derFile);
+
+    return ret == 1;
+}
+
+// Convert DER private key file to PEM format and save to file
+bool derKeyToPem(const std::string& derPath, const std::string& pemPath)
+{
+    FILE* derFile = fopen(derPath.c_str(), "rb");
+    if (!derFile)
+        return false;
+    EVP_PKEYPtr pkey = makeEVPPKeyPtr(d2i_PrivateKey_fp(derFile, nullptr));
+    fclose(derFile);
+    if (!pkey)
+        return false;
+
+    FILE* pemFile = fopen(pemPath.c_str(), "w");
+    if (!pemFile)
+    {
+        return false;
+    }
+    int ret = PEM_write_PrivateKey(pemFile, pkey.get(), nullptr, nullptr, 0,
+                                   nullptr, nullptr);
+    fclose(pemFile);
+    return ret == 1;
+}
+// bool loadTpm2(boost::asio::ssl::context& ctx)
```

Comment:
do you want to keep this tpm code for future use?
you can remove this code for now, will add it when needed